js_modules/dagster-ui/packages/ui-core/src/overview/assets/AssetTimelineDataFetcher.tsx

```diff
@@ -0,0 +1,414 @@
+              },
+            });
+            res(void 0);
+
+            if (isCanceled) {
+              return;
+            }
+
+            const {terminated, unterminated} = data;
+            if (terminated.__typename !== 'Runs' || unterminated.__typename !== 'Runs') {
+              if (terminated.__typename === 'PythonError') {
+                showCustomAlert({
+                  title: 'Error',
+                  body: <PythonErrorInfo error={terminated} />,
+                });
+              }
+              if (unterminated.__typename === 'PythonError') {
+                showCustomAlert({
+                  title: 'Error',
+                  body: <PythonErrorInfo error={unterminated} />,
+                });
+              }
+              return;
+            }
+
+            const groupedRuns = groupRunsByGroup(
+              [...terminated.results, ...unterminated.results],
+              assetToGroup,
+            );
+
+            runsCache.current.locations[locationName] = {
+              runs: [...Array.from(terminated.results), ...Array.from(unterminated.results)],
+              loading: false,
+            };
+            Object.keys(groups).forEach((groupKey) => {
+              const runs = groupedRuns[groupKey]?.runs ?? [];
+              runsCache.current.groups[groupKey] = {
+                runs: Array.from(runs),
+                loading: false,
+              };
+              assetKeys.forEach((key) => {
+                const stringAssetKey = tokenForAssetKey(key);
+                runsCache.current.assets[stringAssetKey] = {
+                  runs: groupedRuns[groupKey]?.assets[stringAssetKey] ?? [],
+                  loading: false,
+                };
+              });
+            });
+            triggerListeners();
+          });
+        }),
+      );
+      isInProgress = false;
+    }
+
+    const interval = setInterval(tick, POLL_INTERVAL);
+    setTimeout(() => tick(true), 100);
+    return () => {
+      isCanceled = true;
+      clearInterval(interval);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [client, didCodeLocationNodesChange, didRangeMeaningfullyChange]);
+
+  const value = React.useMemo(() => {
+    return {
+      subscribeToAssetData: (key: AssetKeyInput, listener: Listener) => {
+        const stringKey = tokenForAssetKey(key);
+        assetSubscriptions.current[stringKey] = listener;
+        const data = runsCache.current.assets[stringKey];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete assetSubscriptions.current[stringKey];
+        };
+      },
+      subscribeToGroupData: (groupName: string, listener: Listener) => {
+        groupSubscriptions.current[groupName] = listener;
+        const data = runsCache.current.groups[groupName];
+        if (data) {
+          listener(data);
+        }
+        return () => {
+          delete groupSubscriptions.current[groupName];
+        };
+      },
```

Comment:
This may be a silly nit, but this breaks if you do something like this:

render 1:
- attach subscriber A

render 2:
- attach subscriber B
- unmount subscriber A


because the unmount removes the subscriber for B.

I think maybe we could say:

```
return () => { 
if (groupSubscriptions.current[groupName] === listener) {
    delete groupSubscriptions.current[groupName];
}
};
```

to safeguard it a bit!

Reply:
Yeah this was not generalized to handle that case because at least for now there will always be exactly 1 subscriber.